Compiler crashed with java.lang.IllegalStateException

Hello,
Today I pulled changes from master branch and bumped to exception when tried to compile my project. I've tried to skip some commits and found that [011197c](https://github.com/JetBrains/kotlin-native/commit/011197c462eceae2a84ae6c751cd0f7169320626) works fine.
Last broken commit that I've tested was [27ca482](https://github.com/JetBrains/kotlin-native/commit/27ca482dabd7dc475389207389c539ccd7b0841b)
I'm not sure, but it seems like [610c9b5](https://github.com/JetBrains/kotlin-native/commit/610c9b52787adca51c48995f3a1e32fcea85d7e6) could be a cause of the issue.

Exception message is completely incomprehensible:
```
 * Compiler version info: Konan: EAP 0.6.1.0 / Kotlin: 1.2.40
 * Output kind: FRAMEWORK

exception: java.lang.IllegalStateException: Symbol for public final fun `$<bridge-UN>next()ValueType`(): kotlin.Any? defined in kotlin.collections.CharIterator[SimpleFunctionDescriptorImpl@c6642c8] is unbound
	at org.jetbrains.kotlin.ir.symbols.impl.IrBindableSymbolBase.getOwner(IrSymbolBase.kt:42)
	at org.jetbrains.kotlin.backend.konan.ir.KonanIr.get(Ir.kt:59)
	at org.jetbrains.kotlin.backend.konan.llvm.objcexport.ObjCExportCodeGeneratorKt.generateObjCImp(ObjCExportCodeGenerator.kt:500)
	at org.jetbrains.kotlin.backend.konan.llvm.objcexport.ObjCExportCodeGeneratorKt.generateObjCImp$default(ObjCExportCodeGenerator.kt:464)
	at org.jetbrains.kotlin.backend.konan.llvm.objcexport.ObjCExportCodeGeneratorKt.createMethodAdapter(ObjCExportCodeGenerator.kt:671)
	at org.jetbrains.kotlin.backend.konan.llvm.objcexport.ObjCExportCodeGeneratorKt.createDirectAdapters(ObjCExportCodeGenerator.kt:886)
	at org.jetbrains.kotlin.backend.konan.llvm.objcexport.ObjCExportCodeGeneratorKt.createTypeAdapter(ObjCExportCodeGenerator.kt:747)
	at org.jetbrains.kotlin.backend.konan.llvm.objcexport.ObjCExportCodeGeneratorKt.access$createTypeAdapter(ObjCExportCodeGenerator.kt:1)
	at org.jetbrains.kotlin.backend.konan.llvm.objcexport.ObjCExportCodeGenerator.emitRtti$backend_native_compiler(ObjCExportCodeGenerator.kt:168)
	at org.jetbrains.kotlin.backend.konan.objcexport.ObjCExport.produceObjCFramework$backend_native_compiler(ObjCExport.kt:84)
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitModuleFragment(IrToBitcode.kt:343)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitModuleFragment(IrElementVisitorVoid.kt:28)
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitModuleFragment(IrToBitcode.kt:253)
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitModuleFragment(IrToBitcode.kt:253)
	at org.jetbrains.kotlin.ir.declarations.impl.IrModuleFragmentImpl.accept(IrModuleFragmentImpl.kt:46)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptVoid(IrElementVisitorVoid.kt:241)
	at org.jetbrains.kotlin.backend.konan.llvm.IrToBitcodeKt$emitLLVM$7.invoke(IrToBitcode.kt:120)
	at org.jetbrains.kotlin.backend.konan.llvm.IrToBitcodeKt$emitLLVM$7.invoke(IrToBitcode.kt)
	at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:140)
	at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:119)
	at org.jetbrains.kotlin.backend.konan.util.UtilKt.profileIf(util.kt:34)
	at org.jetbrains.kotlin.backend.konan.PhaseManager.phase$backend_native_compiler(KonanPhases.kt:139)
	at org.jetbrains.kotlin.backend.konan.llvm.IrToBitcodeKt.emitLLVM(IrToBitcode.kt:119)
	at org.jetbrains.kotlin.backend.konan.KonanDriverKt$runTopLevelPhases$4$2.invoke(KonanDriver.kt:97)
	at org.jetbrains.kotlin.backend.konan.KonanDriverKt$runTopLevelPhases$4$2.invoke(KonanDriver.kt)
	at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:140)
	at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:119)
	at org.jetbrains.kotlin.backend.konan.util.UtilKt.profileIf(util.kt:34)
	at org.jetbrains.kotlin.backend.konan.PhaseManager.phase$backend_native_compiler(KonanPhases.kt:139)
	at org.jetbrains.kotlin.backend.konan.KonanDriverKt$runTopLevelPhases$4.invoke(KonanDriver.kt:96)
	at org.jetbrains.kotlin.backend.konan.KonanDriverKt$runTopLevelPhases$4.invoke(KonanDriver.kt)
	at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:140)
	at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:119)
	at org.jetbrains.kotlin.backend.konan.util.UtilKt.profileIf(util.kt:34)
	at org.jetbrains.kotlin.backend.konan.PhaseManager.phase$backend_native_compiler(KonanPhases.kt:139)
	at org.jetbrains.kotlin.backend.konan.KonanDriverKt.runTopLevelPhases(KonanDriver.kt:90)
	at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:61)
	at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:39)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.java:108)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.java:52)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:92)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:70)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:36)
	at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMainNoExit(CLITool.kt:157)
	at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMain(CLITool.kt:148)
	at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:183)
	at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:174)
	at org.jetbrains.kotlin.backend.konan.util.UtilKt.profileIf(util.kt:34)
	at org.jetbrains.kotlin.backend.konan.util.UtilKt.profile(util.kt:29)
	at org.jetbrains.kotlin.cli.bc.K2Native$Companion.main(K2Native.kt:176)
	at org.jetbrains.kotlin.cli.bc.K2NativeKt.main(K2Native.kt:188)
	at org.jetbrains.kotlin.cli.utilities.MainKt.main(main.kt:27)

 FAILED

FAILURE: Build failed with an exception.
```

--- a/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt
+++ b/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt
@@ -27,6 +27,7 @@ import org.jetbrains.kotlin.backend.konan.llvm.objc.ObjCCodeGenerator
 import org.jetbrains.kotlin.backend.konan.objcexport.*
 import org.jetbrains.kotlin.descriptors.*
 import org.jetbrains.kotlin.ir.declarations.IrConstructor
+import org.jetbrains.kotlin.ir.declarations.IrFunction
 import org.jetbrains.kotlin.ir.declarations.IrSimpleFunction
 import org.jetbrains.kotlin.name.FqName
 import org.jetbrains.kotlin.resolve.descriptorUtil.getSuperClassOrAny
@@ -459,7 +460,7 @@ private fun ObjCExportCodeGenerator.emitSpecialClassesConvertions() {
 }
 
 private fun ObjCExportCodeGenerator.generateObjCImp(
-        target: FunctionDescriptor?,
+        target: IrFunction?,
         methodBridge: MethodBridge,
         isVirtual: Boolean = false
 ): LLVMValueRef {
@@ -497,16 +498,15 @@ private fun ObjCExportCodeGenerator.generateObjCImp(
             }
         }
 
-        val targetIr = context.ir.get(target)
         val llvmTarget = if (!isVirtual) {
-            codegen.llvmFunction(targetIr)
+            codegen.llvmFunction(target)
         } else {
-            lookupVirtualImpl(args.first(), targetIr)
+            lookupVirtualImpl(args.first(), target)
         }
 
         val targetResult = callFromBridge(llvmTarget, args, Lifetime.ARGUMENT)
 
-        if (target is ConstructorDescriptor) {
+        if (target is IrConstructor) {
             ret(param(0))
         } else when (returnType) {
             VoidBridge -> ret(null)
@@ -652,14 +652,14 @@ private fun ObjCExportCodeGenerator.createMethodVirtualAdapter(
     val selector = namer.getSelector(baseMethod)
 
     val methodBridge = mapper.bridgeMethod(baseMethod)
-    val objCToKotlin = constPointer(generateObjCImp(baseMethod, methodBridge, isVirtual = true))
+    val objCToKotlin = constPointer(generateObjCImp(context.ir.get(baseMethod), methodBridge, isVirtual = true))
     return ObjCToKotlinMethodAdapter(selector, getEncoding(methodBridge), objCToKotlin)
 }
 
 private fun ObjCExportCodeGenerator.createMethodAdapter(
         implementation: FunctionDescriptor?,
         baseMethod: FunctionDescriptor
-) = createMethodAdapter(DirectAdapterRequest(implementation, baseMethod))
+) = createMethodAdapter(DirectAdapterRequest(implementation?.let { context.ir.get(it) }, baseMethod))
 
 private fun ObjCExportCodeGenerator.createMethodAdapter(
         request: DirectAdapterRequest
@@ -856,7 +856,7 @@ private fun ObjCExportCodeGenerator.createTypeAdapter(
     )
 }
 
-internal data class DirectAdapterRequest(val implementation: FunctionDescriptor?, val base: FunctionDescriptor)
+internal data class DirectAdapterRequest(val implementation: IrFunction?, val base: FunctionDescriptor)
 
 private fun ObjCExportCodeGenerator.createDirectAdapters(
         method: FunctionDescriptor
@@ -871,7 +871,7 @@ private fun ObjCExportCodeGenerator.createDirectAdapters(
                     context.ir.get(base) as IrSimpleFunction
             ).getImplementation(context)
         }
-        DirectAdapterRequest(implementation?.descriptor, base)
+        DirectAdapterRequest(implementation, base)
     }
 
     val superClassMethod = method.overriddenDescriptors

